From fac79639df82d03c2740d6ecd94eaf75b1d24d43 Mon Sep 17 00:00:00 2001
From: Matt Selsky <matthew.selsky@twosigma.com>
Date: Tue, 6 Dec 2016 01:52:00 -0500
Subject: [PATCH 193/268] Fix update ntp.util.rfc3339 so that all tests pass

Treat the fractional seconds as a string so that we don't change the precision
---
 pylib/util.py            | 9 ++++++---
 tests/pylib/test_util.py | 1 +
 2 files changed, 7 insertions(+), 3 deletions(-)

diff --git a/pylib/util.py b/pylib/util.py
index 5df7bec..360c98a 100644
--- a/pylib/util.py
+++ b/pylib/util.py
@@ -22,10 +22,13 @@ def stdversion():
 
 def rfc3339(t):
     "RFC 3339 string from Unix time, including fractional second."
-    subsec = t - int(t)
-    t -= subsec
     rep = time.strftime("%Y-%m-%dT%H:%M:%S", time.gmtime(t))
-    rep += ("%f" % subsec)[1:] + "Z"
+    t = str(t)
+    if "." in t:
+        subsec = t.split(".", 1)[1]
+        if int(subsec) > 0:
+            rep += "." + subsec
+    rep += "Z"
     return rep
 
 def portsplit(hostname):
diff --git a/tests/pylib/test_util.py b/tests/pylib/test_util.py
index f9aa24d..330a39b 100644
--- a/tests/pylib/test_util.py
+++ b/tests/pylib/test_util.py
@@ -6,6 +6,7 @@ class TestPylibUtilMethods(unittest.TestCase):
     def test_rfc3339(self):
         self.assertEqual(ntp.util.rfc3339(1480999786), '2016-12-06T04:49:46Z')
         self.assertEqual(ntp.util.rfc3339(1480999786.5), '2016-12-06T04:49:46.5Z')
+        self.assertEqual(ntp.util.rfc3339(1480999786.025), '2016-12-06T04:49:46.025Z')
 
     def test_portsplit(self):
         self.assertEqual(ntp.util.portsplit("host.invalid"), ("host.invalid", ""))
-- 
2.7.4

